File: ios/CatholicSaints/Models/Saint.swift
import Foundation

/// Represents a Catholic saint loaded from per-language JSON files.
/// Each language has its own file (saints-en.json, saints-es.json).
struct Saint: Codable, Identifiable, Hashable, Sendable {
    let id: String
    let name: String
    let feastDay: String
    let birthDate: String?
    let deathDate: String?
    let canonizationDate: String?
    let country: String?
    let region: String?
    let gender: String?
    let lifeState: String?
    let ageCategory: String?
    let patronOf: [String]
    let tags: [String]
    let affinities: [String]
    let quote: String?
    let biography: String
    let whyConfirmationSaint: String?
    let image: SaintImage?
    let sources: [String]
    let sourceURLs: [String: String]?

    /// Formatted feast day for display (e.g. "October 1")
    var formattedFeastDay: String {
        let parts = feastDay.split(separator: "-")
        guard parts.count == 2,
              let month = Int(parts[0]),
              let day = Int(parts[1]) else { return feastDay }
        let months = [
            1: "January", 2: "February", 3: "March", 4: "April",
            5: "May", 6: "June", 7: "July", 8: "August",
            9: "September", 10: "October", 11: "November", 12: "December"
        ]
        return "\(months[month] ?? "") \(day)"
    }

    var isYoung: Bool { ageCategory == "young" }
}

struct SaintImage: Codable, Hashable, Sendable {
    let filename: String
    let attribution: String
}

/// Wrapper for decoding the saints JSON file.
struct SaintsFile: Codable, Sendable {
    let version: String
    let language: String
    let lastUpdated: String
    let saints: [Saint]
}

// MARK: - Confirmation Info

struct ConfirmationInfoFile: Codable, Sendable {
    let version: String
    let language: String
    let lastUpdated: String
    let sections: [ConfirmationSection]
}

struct ConfirmationSection: Codable, Identifiable, Hashable, Sendable {
    let id: String
    let title: String
    let content: [ConfirmationContent]
    let sources: [String]
}

struct ConfirmationContent: Codable, Hashable, Sendable {
    let heading: String
    let body: String
}

// MARK: - Categories

struct CategoriesFile: Codable, Sendable {
    let version: String
    let language: String
    let lastUpdated: String
    let categories: [CategoryGroup]
}

struct CategoryGroup: Codable, Identifiable, Hashable, Sendable {
    let id: String
    let name: String
    let description: String
    let icon: String
    let values: [CategoryValue]
}

struct CategoryValue: Codable, Identifiable, Hashable, Sendable {
    let id: String
    let label: String
}

File: android/app/src/main/java/com/yortch/confirmationsaints/data/model/Saint.kt
package com.yortch.confirmationsaints.data.model

import kotlinx.serialization.Serializable

/**
 * A Catholic saint loaded from per-language JSON (saints-en.json / saints-es.json).
 *
 * Mirrors the iOS `Saint` struct (see ios/CatholicSaints/Models/Saint.swift).
 *
 * Cross-platform contract (see docs/android-architecture.md §Appendix B):
 * - Canonical **English** ids drive matching: [patronOf], [tags], [affinities],
 *   [region], [lifeState], [ageCategory], [gender]. Values are identical in both
 *   language files.
 * - Display equivalents for localized rendering live in [displayPatronOf],
 *   [displayTags], [displayAffinities] (Spanish file only; may be null).
 * - [feastDay] is `"MM-DD"`. Formatting to "October 1" / "1 de octubre" is a UI concern.
 */
@Serializable
data class Saint(
    val id: String,
    val name: String,
    val feastDay: String,
    val birthDate: String? = null,
    val deathDate: String? = null,
    val canonizationDate: String? = null,
    val country: String? = null,
    val region: String? = null,
    val gender: String? = null,
    val lifeState: String? = null,
    val ageCategory: String? = null,
    val patronOf: List<String> = emptyList(),
    val displayPatronOf: List<String>? = null,
    val tags: List<String> = emptyList(),
    val displayTags: List<String>? = null,
    val affinities: List<String> = emptyList(),
    val displayAffinities: List<String>? = null,
    val quote: String? = null,
    val biography: String,
    val whyConfirmationSaint: String? = null,
    val image: SaintImage? = null,
    val sources: List<SourceEntry> = emptyList(),
) {
    val isYoung: Boolean get() = ageCategory == "young"
}

@Serializable
data class SaintImage(
    val filename: String,
    val attribution: String,
)

@Serializable
data class SourceEntry(
    val name: String,
    val url: String,
)
